File: sdk-node/src/contract.ts
// Copyright 2026 The Sokel Authors
// SPDX-License-Identifier: Apache-2.0

/**
 * The runtime view of a contract (wire protocol §5).
 *
 * A contract is **data**: `sokel-gen` renders a CONTRACT constant from manifest.yml, and the runtime
 * only looks things up in it and reports it. So there is no second Field validator here — that
 * would be a second definition of the contract, and two definitions drift.
 */

/** One input/output field (protocol §5's Field). Shape only; validation lives in sokel-gen. */
export interface Field {
  name: string;
  label?: string;
  type: string;
  types?: string[];
  required?: boolean;
  default?: unknown;
  desc?: string;
  options?: Array<string | { value: string; label?: string }>;
  fields?: Field[];
  valueType?: Field;
  itemType?: string;
  goType?: string;
  opaque?: boolean;
  oneOf?: Array<{ name: string; label?: string; type: string; fields?: Field[] }>;
}

export interface OperationSpec {
  id: string;
  label?: string;
  desc?: string;
  stream?: boolean;
  internal?: boolean;
  timeoutSec?: number;
  /** Which platform capability slot this operation fills ("rowstore.query" comes from
   * `implements:` in manifest.yml). Absent on ordinary operations. It is reported as-is; the
   * platform routes on it. Missing it here made every generated shell that declares `implements:`
   * fail to compile. */
  capability?: string;
  inputs: Field[];
  outputs: Field[];
}

export interface EventSpec {
  id: string;
  label?: string;
  desc?: string;
  fields: Field[];
}

export interface AuthFlowSpec {
  kind: "qr" | "input" | "oauth";
  steps?: string[];
}

/** The shape of the generated CONTRACT. Keys match the registration payload (protocol §3)
 * verbatim, so they are reported as-is with no translation step. */
export interface ContractData {
  name?: string;
  /** Publisher identity: `<org>/<name>` is how the plugin is addressed in the marketplace. */
  org?: string;
  label?: string;
  desc?: string;
  version?: string;
  operations: OperationSpec[];
  credential_schema?: Field[];
  events?: EventSpec[];
  events_common?: Field[];
  auth_flow?: AuthFlowSpec;
  oauth?: { provider: string; scopes?: string[] };
  capabilities?: Record<string, boolean>;
  /** Translations of the human-facing strings, keyed by locale then by the source string
   * (`{"zh-CN": {"Row query": "按行查询"}}`). Generated from `locales/<lang>.json` next to the
   * manifest; the platform renders whichever locale the viewer is in. */
  locales?: Record<string, Record<string, string>>;
  doc?: string;
  doc_url?: string;
}

/** Reserved operation ids (the auth flow).
 *
 * A **plain** business id cannot contain a dot (it must match ^[a-z][a-z0-9_]*$), but a capability
 * slot's id does — `implements:` produces "rowstore.query" — so "has a dot" no longer means
 * "reserved". Compare against these constants rather than looking for a dot. */
export const OP_AUTH_START = "auth.start";
export const OP_AUTH_POLL = "auth.poll";
export const OP_AUTH_SUBMIT = "auth.submit";

/** The special operation name for platform-relayed webhooks (it reuses the call frame, §7b). */
export const OP_WEBHOOK = "__webhook__";

/** Capability bit: registering a webhook handler *is* the declaration; the author does not repeat it. */
export const CAP_WEBHOOK = "webhook";

export class Contract {
  constructor(readonly data: ContractData) {}

  operations(): OperationSpec[] {
    return this.data.operations ?? [];
  }

  operation(id: string): OperationSpec | undefined {
    return this.operations().find((op) => op.id === id);
  }

  isStream(id: string): boolean {
    return this.operation(id)?.stream === true;
  }

  eventIds(): string[] {
    return (this.data.events ?? []).map((e) => e.id);
  }

  /** The contract half of the registration payload. Empty values are omitted (the protocol makes
   * every new field optional). */
  payload(): Record<string, unknown> {
    const out: Record<string, unknown> = { operations: this.operations() };
    const d = this.data;
    if (d.credential_schema?.length) out.credential_schema = d.credential_schema;
    if (d.events?.length) out.events = d.events;
    if (d.events_common?.length) out.events_common = d.events_common;
    if (d.auth_flow) out.auth_flow = d.auth_flow;
    if (d.oauth) out.oauth = d.oauth;
    if (d.capabilities && Object.keys(d.capabilities).length) out.capabilities = d.capabilities;
    if (d.doc) out.doc = d.doc;
    if (d.doc_url) out.doc_url = d.doc_url;
    return out;
  }
}

File: sdk-python/sokel/contract.py
"""契约的运行时视图（线协议 §5）。

契约本身是**数据**：`sokel-gen` 从 sokel.yaml 生成一份 CONTRACT 字典，运行时只是查它、上报它。
所以这里不重新定义一套 Field 类——那会变成契约的第二份定义，而两份定义迟早会漂
（Go 侧栽过一次：SDK 的 Field 是全量的、平台那份只有四个键，SDK 声明了的东西平台看不见）。
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

# 契约字典的键（与线协议 §3 的注册载荷同名，直接上报，不做转换）
KEY_OPERATIONS = "operations"
KEY_CREDENTIAL = "credential_schema"
KEY_EVENTS = "events"
KEY_EVENTS_COMMON = "events_common"
KEY_AUTH_FLOW = "auth_flow"
KEY_OAUTH = "oauth"
KEY_CAPABILITIES = "capabilities"
KEY_DOC = "doc"
KEY_DOC_URL = "doc_url"

# 保留操作 id（认证流）。带点号，业务 id 产生不出来（业务 id 限定 ^[a-z][a-z0-9_]*$）。
OP_AUTH_START = "auth.start"
OP_AUTH_POLL = "auth.poll"
OP_AUTH_SUBMIT = "auth.submit"

# 平台代收 webhook 的特殊操作名（复用调用帧，见协议 §7b）
OP_WEBHOOK = "__webhook__"

# 能力位：注册了 webhook 处理器就是支持，不靠作者手动声明
CAP_WEBHOOK = "webhook"


class Contract:
    """一份插件契约。构造参数就是生成物 CONTRACT 字典。"""

    def __init__(self, data: Optional[Dict[str, Any]] = None) -> None:
        self.data: Dict[str, Any] = dict(data or {})

    # —— 查 ——

    def operations(self) -> List[Dict[str, Any]]:
        return list(self.data.get(KEY_OPERATIONS) or [])

    def operation(self, op_id: str) -> Optional[Dict[str, Any]]:
        for op in self.operations():
            if op.get("id") == op_id:
                return op
        return None

    def is_stream(self, op_id: str) -> bool:
        op = self.operation(op_id)
        return bool(op and op.get("stream"))

    def event_ids(self) -> List[str]:
        return [e.get("id", "") for e in (self.data.get(KEY_EVENTS) or [])]

    def get(self, key: str, default: Any = None) -> Any:
        return self.data.get(key, default)

    def set(self, key: str, value: Any) -> None:
        self.data[key] = value

    # —— 上报 ——

    def payload(self) -> Dict[str, Any]:
        """契约部分的注册载荷。空值一律省略（协议：新字段一律 optional）。"""
        out: Dict[str, Any] = {}
        for key in (
            KEY_OPERATIONS,
            KEY_CREDENTIAL,
            KEY_EVENTS,
            KEY_EVENTS_COMMON,
            KEY_AUTH_FLOW,
            KEY_OAUTH,
            KEY_CAPABILITIES,
            KEY_DOC,
            KEY_DOC_URL,
        ):
            v = self.data.get(key)
            if v:
                out[key] = v
        # operations 必须在（哪怕空）：平台侧读不到这个键会当成「契约锁定」而不是「没有操作」
        out.setdefault(KEY_OPERATIONS, [])
        return out
